### PR TITLE
Calls to stash() inside of persist callbacks should throw exception

### PR DESCRIPTION
**Pull request by ortibazar**
Refs #18470

### CODE REVIEW

**Review by ktoso — APPROVED** [review]
LGTM, nice improvement - thanks!

File: akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala
Comment (patriknw) [inline akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala]: This requires some thinking regarding binary compatibility. Since this is a trait we can normally not change much of it. My thinking is that it might be alright here. If a library is compiled with an earlier Akka version this is not included in those class files, but the methods using this field are also in the same trait and would also not be the library (well, old versions of the methods would).

**Review by patriknw — COMMENTED** [review]
looking good, but I would like confirmation from some more that we are not breaking bin compat here

File: akka-persistence/src/test/scala/akka/persistence/PersistentActorStashingSpec.scala
Comment (raboof) [inline akka-persistence/src/test/scala/akka/persistence/PersistentActorStashingSpec.scala]: This is a bit of an 'indirect' way of testing - would it be possible to let the actor crash and get the exception from the supervisor?

File: akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala
Comment (raboof) [inline akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala]: Perhaps we could go into some more detail on why this is bad and what to do instead?

**Review by raboof — COMMENTED** [review]
I like the safeguard - perhaps the message and the test could be refined a bit?

File: project/MiMa.scala
Comment (raboof) [inline project/MiMa.scala]: Hmm these need to be as 2.5 has been released since this PR was opened.

Comment (ortigali) [inline project/MiMa.scala]: What should I do? Or Should I wait until 2.6 version? Could you give me some hints?

Comment (raboof) [inline project/MiMa.scala]: Ah, no, just rebase this PR onto lastest master and move these MiMa exceptions to the latest version instead of under 2.4.16 where they are now.

The question whether this PR breaks binary compatiblity is a separate one: if it breaks binary compatibility in a problematic way it'll have to move to akka 3, but it looks like it might not. Would be good to get another pair of eyes on that though.

File: akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala
Comment (patriknw) [inline akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala]: @ktoso do you think this is alright?

Comment (ktoso) [reply akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala]: I'll test it out

Comment (ktoso) [reply akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala]: I'll try to reproduce the exact case since users don't extend Eventsourced but PersistentActor...

But in general the problem is that `private var` still is a public method on the interface generated:

```
$ cat Example.scala
trait Example {
  private var num = 21
  def lol() = ""

}
ktoso @ 三日月/tmp
(reverse-i-search)`ja': java -cp $SCALA_LIB_2_12:. User
ktoso @ 三日月/tmp
$ javap -cp . -p Example
Compiled from "Example.scala"
public interface Example {
  public abstract int Example$$num(); // OH NOES!!!
  public abstract void Example$$num_$eq(int);// OH NOES!!!

  public static java.lang.String lol$(Example);
  public java.lang.String lol();
  public static void $init$(Example);
}
```

So when someone extended it, back then, and now would use it with the new Example on classpath they'd get:

```
$ java -cp $SCALA_LIB_2_12:. User
Exception in thread "main" java.lang.AbstractMethodError: Method User.Example$$num_$eq(I)V is abstract
```

I'll see if our layer of PersistentActor is helping here or not though

Comment (ktoso) [reply akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala]: I'm sad to report that this will blow up in this situation:

```
$ cat Example.scala
trait Example {
  private var num = 21
  def lol() = "hello  "

}
ktoso @ 三日月/tmp
$ cat PA.scala
trait PA extends Example {}
ktoso @ 三日月/tmp
$ echo "PA and Example are compiled together"
PA and Example are compiled together
ktoso @ 三日月/tmp
$
ktoso @ 三日月/tmp
$ echo User.scala
User.scala
ktoso @ 三日月/tmp
$ java -cp $SCALA_LIB_2_12:. User
Exception in thread "main" java.lang.AbstractMethodError: Method User.Example$$num_$eq(I)V is abstract
	at User.Example$$num_$eq(User.scala)
	at Example.$init$(Example.scala:2)
```

So creating an PersistentActor by users (User) would become broken if we do this.

**Review by ktoso — CHANGES_REQUESTED** [review]
Binary incompatibility I believe :-(
Hard to solve this one hm.

File: akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala
Comment (patriknw) [reply akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala]: The current message that is to be stashed is an internal message from the journal confirming the persist (`WriteMessageSuccess`). Stashing that is obviously not desired.

Comment (patriknw) [reply akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala]: We can solve it without the flag. We can check if the currentMessage is
```
context.asInstanceOf[ActorCell].currentMessage match {
  case _: JournalProtocol.Response => throw new IllegalStateException(...
  case _ => super.stash()
}
```

This also has the benefit that we cover some more cases, such as messages during recovery.

**Review by patriknw — CHANGES_REQUESTED** [review]
@ortigali I think we can solve it in another binary compatible way. Could you take a look at this...

File: akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala
Comment (patriknw) [reply akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala]: @ortigali here is the suggestion ^

Comment (patriknw) [reply akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala]: @ortigali Will you be able to complete this with the approach suggested, or should we take over from here?

**Comment by ortigali** [comment]
@patriknw could you review it?

File: akka-persistence/src/main/mima-filters/2.5.6.backwards.excludes
Comment (ktoso) [inline akka-persistence/src/main/mima-filters/2.5.6.backwards.excludes]: hm is this still needed? seems we just added an override which should be fine hm

File: akka-persistence/src/test/scala/akka/persistence/PersistentActorStashingSpec.scala
Comment (ktoso) [reply akka-persistence/src/test/scala/akka/persistence/PersistentActorStashingSpec.scala]: Agree with @raboof; I'd propose to send back the exception to the test rather than the event, and assert on it's message here in the test rather than in the actor

**Review by ktoso — COMMENTED** [review]
Looks good, small comment in-line :)

File: akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala
Comment (patriknw) [inline akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala]: will this catch the case when stash is called during recovery also? then the err msg should "... or during recovery"

File: akka-persistence/src/test/scala/akka/persistence/PersistentActorStashingSpec.scala
Comment (patriknw) [reply akka-persistence/src/test/scala/akka/persistence/PersistentActorStashingSpec.scala]: 👍

**Review by patriknw — COMMENTED** [review]
excellent, just some small comments

File: akka-persistence/src/test/scala/akka/persistence/PersistentActorStashingSpec.scala
Comment (ortigali) [reply akka-persistence/src/test/scala/akka/persistence/PersistentActorStashingSpec.scala]: @ktoso "send back the exception to the test rather than the event" how can i return it back to the test. You meant via supervisor

File: akka-persistence/src/main/mima-filters/2.5.6.backwards.excludes
Comment (ktoso) [reply akka-persistence/src/main/mima-filters/2.5.6.backwards.excludes]: Ping about this one, still needed?

File: akka-persistence/src/test/scala/akka/persistence/PersistentActorStashingSpec.scala
Comment (ktoso) [inline akka-persistence/src/test/scala/akka/persistence/PersistentActorStashingSpec.scala]: call them something like "stash-in-persist" etc?

File: akka-persistence/src/main/mima-filters/2.5.6.backwards.excludes
Comment (ortigali) [reply akka-persistence/src/main/mima-filters/2.5.6.backwards.excludes]: it is needed, mima report says
`synthetic method akka$persistence$Eventsourced$$super$stash()Unit in trait akka.persistence.Eventsourced is present only in current version`

**Review by ktoso — APPROVED** [review]
Looking good, small question

File: akka-persistence/src/main/mima-filters/2.5.6.backwards.excludes
Comment (patriknw) [reply akka-persistence/src/main/mima-filters/2.5.6.backwards.excludes]: @ortigali sorry for review delays, this should now be in 2.5.7.backwards.excludes

**Review by patriknw — APPROVED** [review]
LGTM

**Comment by patriknw** [comment]
i fixed the merge conflict, let's squash-merge this when validated again

**Comment by patriknw** [comment]
sbt NPE, I'm so tired of that one

**Comment by patriknw** [comment]
PLS BUILD

**Comment by patriknw** [comment]
PLS BUILD

**Comment by patriknw** [comment]
Unrelated failure, created https://github.com/akka/akka/issues/24113

**Comment by patriknw** [comment]
PLS BUILD

**Comment by raboof** [comment]
failure is #23859 (which seems really rampant today :( )

**Comment by raboof** [comment]
PLS BUILD

**Comment by raboof** [comment]
again

**Comment by raboof** [comment]
PLS BUILD

**Comment by raboof** [comment]
PLS BUILD

**Comment by patriknw** [comment]
thanks @ortigali

**Pull request merged**
